src/jot.gleam

```diff
@@ -1277,15 +1447,67 @@ fn take_list_item_chars(
   }
 }
 
-fn take_paragraph_chars(in: String) -> #(String, String) {
-  case string.split_once(in, "\n\n") {
+fn take_paragraph_chars(
+  in: String,
+  div_close_size: Option(Int),
+) -> #(String, String) {
+  let #(paragraph, rest) = case string.split_once(in, "\n\n") {
     Ok(#(content, in)) -> #(content, in)
     Error(_) ->
       case string.ends_with(in, "\n") {
         True -> #(string.drop_end(in, 1), "")
         False -> #(in, "")
       }
   }
+
+  case div_close_size {
+    Some(size) -> {
+      let #(split_paragraph, paragraph_rest) =
+        search_paragraph_for_div_end(paragraph, [], size)
+
+      case split_paragraph, paragraph_rest {
+        "", "" -> #(paragraph, rest)
+        _, "" -> #(split_paragraph, rest)
+        _, _ -> #(split_paragraph, paragraph_rest <> "\n\n" <> rest)
+      }
+    }
+    None -> #(paragraph, rest)
+  }
+}
+
+/// search a stretch of paragraph characters for valid paragraph terminator
+///
+/// regex expression: `\n[ \t]*:::+[ \t]*(\n|$)`
+fn search_paragraph_for_div_end(
+  in: String,
+  acc: List(String),
+  size: Int,
+) -> #(String, String) {
+  let #(line, rest) = slurp_to_line_end(in)
+  case check_line_suitable_div_end(line, size) {
+    True -> {
+      #(acc |> list.reverse |> string.join("\n"), rest)
+    }
+    False -> {
+      case rest {
+        "" -> #([line, ..acc] |> list.reverse |> string.join("\n"), "")
+        rest -> search_paragraph_for_div_end(rest, [line, ..acc], size)
+      }
+    }
+  }
+}
+
+/// Split at \n. if a newline is not present ans there are remaining characters,
+/// then the remaining characters will be returned in the line position with an empty rest
+fn slurp_to_line_end(in: String) -> #(String, String) {
+  let split =
+    string.split_once(in, "\n")
+    |> option.from_result
```

Comment:
Remove this, there's no need to convert to option here as it's immediately matched on below.